<div align="center">

# Investigating NICS and Census Datasets

</div>

## Abstract


## Table of Contents
<ul>
<li><a href="#intro">Introduction</a>
    <ul>
        <li>Dataset Description</li>
        <li>Questions for analysis</li>
    </ul>
</li>
<li><a href="#wrangling">Data Wrangling</a>
    <ul>
        <li>Loading the datasets</li>
        <li>Initial exploration</li>
        <li>Data assessment
            <ul>
                <li>Missing values</li>
                <li>Duplicated values</li>
                <li>Outliers</li>
                <li>Inconsistencies</li>
            </ul>
        </li>
        <li>Data cleaning</li>
    </ul>
</li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
    <ul>
        <li>Question 1:</li>
        <li>Question 2:</li>
    </ul>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction
In the united State of America,

### Dataset Description 

The dataset used in this analysis comes from the FBI's National Instant Criminal Background Check System. The NICS is used by to determine whether a prospective buyer is eligible to buy firearms or explosives. Gun shops call into this system to ensure that each customer does not have a criminal record or isn’t otherwise ineligible to make a purchase. The data has been supplemented with state level data from census.gov.

- **The NICS data:** is found in .csv file. It contains the number of firearm checks by month, state, and type.
The following table shows the details for each column in the dataset.

<div align="center">

**Table 1:** Columns of NICS data with describtion.

| Column Name | Description |
|---|---|
| month | The month the data applies to. |
| state | The state where the data was collected. |
| permit | Number of permits issued. |
| permit_recheck | Number of permit rechecks conducted. |
| handgun | Number of handgun transactions (could be permit applications, purchases, etc.). |
| long_gun | Number of long gun transactions (could be permit applications, purchases, etc.). |
| other | Number of other firearm transactions (could be permit applications, purchases, etc.). |
| multiple | Number of transactions involving multiple firearm types. |
| admin | Number of administrative transactions (could be background checks, license renewals, etc.). |
| prepawn_handgun | Number of handguns pre-pawned. |
| prepawn_long_gun | Number of long guns pre-pawned. |
| prepawn_other | Number of other firearms pre-pawned. |
| redemption_handgun | Number of handguns redeemed from pawn. |
| redemption_long_gun | Number of long guns redeemed from pawn. |
| redemption_other | Number of other firearms redeemed from pawn. |
| returned_handgun | Number of handguns returned to seller. |
| returned_long_gun | Number of long guns returned to seller. |
| returned_other | Number of other firearms returned to seller. |
| rentals_handgun | Number of handgun rentals. |
| rentals_long_gun | Number of long gun rentals. |
| private_sale_handgun | Number of handguns sold privately. |
| private_sale_long_gun | Number of long guns sold privately. |
| private_sale_other | Number of other firearms sold privately. |
| return_to_seller_handgun | Number of handguns returned by buyer to seller (not pawn). |
| return_to_seller_long_gun | Number of long guns returned by buyer to seller (not pawn). |
| return_to_seller_other | Number of other firearms returned by buyer to seller (not pawn). |
| totals | Total number of transactions. |

</div>

- **The U.S. census data:** is found in a .csv file. It contains several variables at the state level. Most variables just have one data point per state (2016), but a few have data for more than one year. The following table shows the details for each column in the dataset.

<div align="center">

**Table 2:** Columns of US census data with describtion.

| Column Name | Description |
|---|---|
| Fact | The overall category or type of the data. |
| Fact Note | Additional notes or details about the Fact column. |
| Alabama | Data specific to Alabama. |
| Alaska | Data specific to Alaska. |
| Arizona | Data specific to Arizona. |
| Arkansas | Data specific to Arkansas. |
| California | Data specific to California. |
| Colorado | Data specific to Colorado. |
| Connecticut | Data specific to Connecticut. |
| Delaware | Data specific to Delaware. |
| Florida | Data specific to Florida. |
| Georgia | Data specific to Georgia. |
| Hawaii | Data specific to Hawaii. |
| Idaho | Data specific to Idaho. |
| Illinois | Data specific to Illinois. |
| Indiana | Data specific to Indiana. |
| Iowa | Data specific to Iowa. |
| Kansas | Data specific to Kansas. |
| Kentucky | Data specific to Kentucky. |
| Louisiana | Data specific to Louisiana. |
| Maine | Data specific to Maine. |
| Maryland | Data specific to Maryland. |
| Massachusetts | Data specific to Massachusetts. |
| Michigan | Data specific to Michigan. |
| Minnesota | Data specific to Minnesota. |
| Mississippi | Data specific to Mississippi. |
| Missouri | Data specific to Missouri. |
| Montana | Data specific to Montana. |
| Nebraska | Data specific to Nebraska. |
| Nevada | Data specific to Nevada. |
| New Hampshire | Data specific to New Hampshire. |
| New Jersey | Data specific to New Jersey. |
| New Mexico | Data specific to New Mexico. |
| New York | Data specific to New York. |
| North Carolina | Data specific to North Carolina. |
| North Dakota | Data specific to North Dakota. |
| Ohio | Data specific to Ohio. |
| Oklahoma | Data specific to Oklahoma. |
| Oregon | Data specific to Oregon. |
| Pennsylvania | Data specific to Pennsylvania. |
| Rhode Island | Data specific to Rhode Island. |
| South Carolina | Data specific to South Carolina. |
| South Dakota | Data specific to South Dakota. |
| Tennessee | Data specific to Tennessee. |
| Texas | Data specific to Texas. |
| Utah | Data specific to Utah. |
| Vermont | Data specific to Vermont. |
| Virginia | Data specific to Virginia. |
| Washington | Data specific to Washington. |
| West Virginia | Data specific to West Virginia. |
| Wisconsin | Data specific to Wisconsin. |
| Wyoming | Data specific to Wyoming. |

</div>

### Question(s) for Analysis
- state one or more questions that you plan on exploring over the course of the report.
- address these questions in the **data analysis** and **conclusion** sections. 
- Try to build your report around the analysis of at least one dependent variable and three independent variables. 
- If you're not sure what questions to ask, then make sure you familiarize yourself with the dataset, its variables and the dataset context for ideas of what to explore.

Example from them:
- What census data is most associated with high gun per capita? 
- Which states have had the highest growth in gun registrations? 
- What is the overall trend of gun purchases?

Necessary imports:

In [13]:
import pandas as pd
import numpy as np
import os
from IPython.display import display

<a id='wrangling'></a>
## Data Wrangling
In this part of the project, the data wrangling pipeline is implemented. Here is the subsections of this pipeline:
- **Loading in the data**: reading data files into pandas dataframes.
- **Initial exploration**: looking into the head, tail, and info of each dataset.
- **Assessing the data**: missing data, duplicated data, outliers, and inconsistent data between the two datasets.
- **Data Cleaning**: decisions based on the assessment to trim and clean the datasets for analysis and finally merge them into a single dataset.

### **Loading the datasets**

In [257]:
def load_data(path="./Database_Ncis_and_Census_data"):
    """
    This function loads the dataset into pandas dataFrames
    Args:
        path(str): Path to the datasets files with the deafult value
    Returns:
        dfs (list): list of of pandas dataFrames (pd.DataFrame)
    """
    dfs = []
    for file in os.listdir(path):
        if file.endswith(".csv"):
            dfs.append(pd.read_csv(path+'/'+file))
        else:
            print(f"{file} is not a csv file")
    return dfs

In [258]:
dfs = load_data()
guns_dataset = dfs[0]
census_dataset = dfs[1]

### **Initial Exploration the Datasets**
This will include looking into the following:
- Head, Tail, Info
- Statics Description
- Number of Unique Values

In [259]:
def initial_exploration(df):
    """
    This function explores the dataset with missing, info, and outliers
    Args:
        df(pd.DataFrame):
    Returns:
        Nothing, displays the dataset explorations
    """
    if isinstance(df, pd.DataFrame):
        print("Shape of this dataset is: ", df.shape)
        display(df.head())
        display(df.tail())
        display(df.info())
    else:
        print("TypeError: This is not a pandas DataFrame")

#### Guns Dataset:
First of all let's explore the guns dataset:

In [260]:
initial_exploration(guns_dataset)

Shape of this dataset is:  (12485, 27)


month       state   permit  permit_recheck  handgun  long_gun   other  \
0  2017-09     Alabama  16717.0             0.0   5734.0    6320.0   221.0   
1  2017-09      Alaska    209.0             2.0   2320.0    2930.0   219.0   
2  2017-09     Arizona   5069.0           382.0  11063.0    7946.0   920.0   
3  2017-09    Arkansas   2935.0           632.0   4347.0    6063.0   165.0   
4  2017-09  California  57839.0             0.0  37165.0   24581.0  2984.0   

   multiple  admin  prepawn_handgun  ...  returned_other  rentals_handgun  \
0       317    0.0             15.0  ...             0.0              0.0   
1       160    0.0              5.0  ...             0.0              0.0   
2       631    0.0             13.0  ...             0.0              0.0   
3       366   51.0             12.0  ...             0.0              0.0   
4         0    0.0              0.0  ...             0.0              0.0   

   rentals_long_gun  private_sale_handgun  private_sale_long_gun  \
0               0.0                   9.0                   16.0   
1               0.0                  17.0                   24.0   
2               0.0                  38.0                   12.0   
3               0.0                  13.0                   23.0   
4               0.0                   0.0                    0.0   

   private_sale_other  return_to_seller_handgun  return_to_seller_long_gun  \
0                 3.0                       0.0                        0.0   
1                 1.0                       0.0                        0.0   
2                 2.0                       0.0                        0.0   
3                 0.0                       0.0                        2.0   
4                 0.0                       0.0                        0.0   

   return_to_seller_other  totals  
0                     3.0   32019  
1                     0.0    6303  
2                     0.0   28394  
3                     1.0   17747  
4                     0.0  123506  

[5 rows x 27 columns]

month          state  permit  permit_recheck  handgun  long_gun  \
12480  1998-11       Virginia     0.0             NaN     14.0       2.0   
12481  1998-11     Washington     1.0             NaN     65.0     286.0   
12482  1998-11  West Virginia     3.0             NaN    149.0     251.0   
12483  1998-11      Wisconsin     0.0             NaN     25.0     214.0   
12484  1998-11        Wyoming     8.0             NaN     45.0      49.0   

       other  multiple  admin  prepawn_handgun  ...  returned_other  \
12480    NaN         8    0.0              NaN  ...             NaN   
12481    NaN         8    1.0              NaN  ...             NaN   
12482    NaN         5    0.0              NaN  ...             NaN   
12483    NaN         2    0.0              NaN  ...             NaN   
12484    NaN         5    0.0              NaN  ...             NaN   

       rentals_handgun  rentals_long_gun  private_sale_handgun  \
12480              NaN               NaN                   NaN   
12481              NaN               NaN                   NaN   
12482              NaN               NaN                   NaN   
12483              NaN               NaN                   NaN   
12484              NaN               NaN                   NaN   

       private_sale_long_gun  private_sale_other  return_to_seller_handgun  \
12480                    NaN                 NaN                       NaN   
12481                    NaN                 NaN                       NaN   
12482                    NaN                 NaN                       NaN   
12483                    NaN                 NaN                       NaN   
12484                    NaN                 NaN                       NaN   

       return_to_seller_long_gun  return_to_seller_other  totals  
12480                        NaN                     NaN      24  
12481                        NaN                     NaN     361  
12482                        NaN                     NaN     408  
12483                        NaN                     NaN     241  
12484                        NaN                     NaN     107  

[5 rows x 27 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12485 entries, 0 to 12484
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   month                      12485 non-null  object 
 1   state                      12485 non-null  object 
 2   permit                     12461 non-null  float64
 3   permit_recheck             1100 non-null   float64
 4   handgun                    12465 non-null  float64
 5   long_gun                   12466 non-null  float64
 6   other                      5500 non-null   float64
 7   multiple                   12485 non-null  int64  
 8   admin                      12462 non-null  float64
 9   prepawn_handgun            10542 non-null  float64
 10  prepawn_long_gun           10540 non-null  float64
 11  prepawn_other              5115 non-null   float64
 12  redemption_handgun         10545 non-null  float64
 13  redemption_long_gun        10544 non-null  flo

None

>This dataset (Gun dataset) has the Shape of (12485, 27), it contains 12485 instances and 27 features. These features are mainly numerical of types such as int, float. It is noticted that these fetaurs have various features scales given that the means of each is on diffrent scales [0-1, 0-12, 0-71, 0-200...] etc. From the initial look, the non-null value counts for each column has diffrent values than the shape of the data which has to be (12485), that indicates the missing values. Another indication for the missing valeus, is when looking at the tail of the datset, the NaN values are observed. For the memory usage of this dataset it is 2.6+ MB

- Let's look at the unique values for each column to later make better datatypes optimizations:

In [263]:
guns_dataset.nunique()

month                          227
state                           55
permit                        5390
permit_recheck                 168
handgun                       7381
long_gun                      8350
other                         1226
multiple                      1387
admin                          499
prepawn_handgun                 90
prepawn_long_gun               133
prepawn_other                   16
redemption_handgun            1893
redemption_long_gun           2370
redemption_other                47
returned_handgun               237
returned_long_gun              113
returned_other                  34
rentals_handgun                  9
rentals_long_gun                 8
private_sale_handgun           152
private_sale_long_gun          136
private_sale_other              43
return_to_seller_handgun        17
return_to_seller_long_gun       17
return_to_seller_other           5
totals                       10218
dtype: int64

> It is noticed that

- Statistics Descriptions for each column, which is permutated for better readability:

In [274]:
guns_dataset.describe().T

count          mean           std  min      25%  \
permit                     12461.0   6413.629404  23752.338269  0.0     0.00   
permit_recheck              1100.0   1165.956364   9224.200609  0.0     0.00   
handgun                    12465.0   5940.881107   8618.584060  0.0   865.00   
long_gun                   12466.0   7810.847585   9309.846140  0.0  2078.25   
other                       5500.0    360.471636   1349.478273  0.0    17.00   
multiple                   12485.0    268.603364    783.185073  0.0    15.00   
admin                      12462.0     58.898090    604.814818  0.0     0.00   
prepawn_handgun            10542.0      4.828021     10.907756  0.0     0.00   
prepawn_long_gun           10540.0      7.834156     16.468028  0.0     0.00   
prepawn_other               5115.0      0.165591      1.057105  0.0     0.00   
redemption_handgun         10545.0    407.970413    783.467053  0.0     0.00   
redemption_long_gun        10544.0    599.332417    952.144809  0.0     0.00   
redemption_other            5115.0      1.815249      4.574641  0.0     0.00   
returned_handgun            2200.0     29.612727     81.943020  0.0     0.00   
returned_long_gun           2145.0      7.557576     22.097046  0.0     0.00   
returned_other              1815.0      1.027548      4.386296  0.0     0.00   
rentals_handgun              990.0      0.076768      0.634503  0.0     0.00   
rentals_long_gun             825.0      0.087273      0.671649  0.0     0.00   
private_sale_handgun        2750.0     14.936000     71.216021  0.0     0.00   
private_sale_long_gun       2750.0     11.602909     54.253090  0.0     0.00   
private_sale_other          2750.0      1.030182      4.467843  0.0     0.00   
return_to_seller_handgun    2475.0      0.402020      1.446568  0.0     0.00   
return_to_seller_long_gun   2750.0      0.441818      1.528223  0.0     0.00   
return_to_seller_other      2255.0      0.105987      0.427363  0.0     0.00   
totals                     12485.0  21595.725911  32591.418387  0.0  4638.00   

                               50%       75%       max  
permit                       518.0   4272.00  522188.0  
permit_recheck                 0.0      0.00  116681.0  
handgun                     3059.0   7280.00  107224.0  
long_gun                    5122.0  10380.75  108058.0  
other                        121.0    354.00   77929.0  
multiple                     125.0    301.00   38907.0  
admin                          0.0      0.00   28083.0  
prepawn_handgun                0.0      5.00     164.0  
prepawn_long_gun               1.0      8.00     269.0  
prepawn_other                  0.0      0.00      49.0  
redemption_handgun            39.0    561.00   10046.0  
redemption_long_gun          175.0    872.00    8831.0  
redemption_other               0.0      2.00      79.0  
returned_handgun               0.0     16.00     603.0  
returned_long_gun              0.0      4.00     168.0  
returned_other                 0.0      0.00      64.0  
rentals_handgun                0.0      0.00      12.0  
rentals_long_gun               0.0      0.00      12.0  
private_sale_handgun           0.0      2.00    1017.0  
private_sale_long_gun          0.0      4.00     777.0  
private_sale_other             0.0      0.00      71.0  
return_to_seller_handgun       0.0      0.00      28.0  
return_to_seller_long_gun      0.0      0.00      17.0  
return_to_seller_other         0.0      0.00       4.0  
totals                     12399.0  25453.00  541978.0

> It is noticed that

#### Census Dataset:
Now let's explore the Census dataset:

In [261]:
initial_exploration(census_dataset)

Shape of this dataset is:  (85, 52)


Fact Fact Note    Alabama  \
0       Population estimates, July 1, 2016,  (V2016)       NaN  4,863,300   
1  Population estimates base, April 1, 2010,  (V2...       NaN  4,780,131   
2  Population, percent change - April 1, 2010 (es...       NaN      1.70%   
3                  Population, Census, April 1, 2010       NaN  4,779,736   
4  Persons under 5 years, percent, July 1, 2016, ...       NaN      6.00%   

    Alaska    Arizona   Arkansas  California   Colorado Connecticut Delaware  \
0  741,894  6,931,071  2,988,248  39,250,017  5,540,545   3,576,452  952,065   
1  710,249  6,392,301  2,916,025  37,254,522  5,029,324   3,574,114  897,936   
2    4.50%      8.40%      2.50%       5.40%     10.20%       0.10%    6.00%   
3  710,231  6,392,017  2,915,918  37,253,956  5,029,196   3,574,097  897,934   
4    7.30%      6.30%      6.40%       6.30%      6.10%       5.20%    5.80%   

   ... South Dakota Tennessee       Texas       Utah  Vermont   Virginia  \
0  ...       865454   6651194  27,862,596  3,051,217  624,594  8,411,808   
1  ...       814195   6346298  25,146,100  2,763,888  625,741  8,001,041   
2  ...        0.063     0.048      10.80%     10.40%   -0.20%      5.10%   
3  ...       814180   6346105  25,145,561  2,763,885  625,741  8,001,024   
4  ...        0.071     0.061       7.20%      8.30%    4.90%      6.10%   

  Washington West Virginia  Wisconsin  Wyoming  
0  7,288,000     1,831,102  5,778,708  585,501  
1  6,724,545     1,853,011  5,687,289  563,767  
2      8.40%        -1.20%      1.60%    3.90%  
3  6,724,540     1,852,994  5,686,986  563,626  
4      6.20%         5.50%      5.80%    6.50%  

[5 rows x 52 columns]

Fact                                          Fact Note Alabama Alaska  \
80   FN             Footnote on this item in place of data     NaN    NaN   
81  NaN                                      Not available     NaN    NaN   
82    S    Suppressed; does not meet publication standards     NaN    NaN   
83    X                                     Not applicable     NaN    NaN   
84    Z  Value greater than zero but less than half uni...     NaN    NaN   

   Arizona Arkansas California Colorado Connecticut Delaware  ...  \
80     NaN      NaN        NaN      NaN         NaN      NaN  ...   
81     NaN      NaN        NaN      NaN         NaN      NaN  ...   
82     NaN      NaN        NaN      NaN         NaN      NaN  ...   
83     NaN      NaN        NaN      NaN         NaN      NaN  ...   
84     NaN      NaN        NaN      NaN         NaN      NaN  ...   

   South Dakota Tennessee Texas Utah Vermont Virginia Washington  \
80          NaN       NaN   NaN  NaN     NaN      NaN        NaN   
81          NaN       NaN   NaN  NaN     NaN      NaN        NaN   
82          NaN       NaN   NaN  NaN     NaN      NaN        NaN   
83          NaN       NaN   NaN  NaN     NaN      NaN        NaN   
84          NaN       NaN   NaN  NaN     NaN      NaN        NaN   

   West Virginia Wisconsin Wyoming  
80           NaN       NaN     NaN  
81           NaN       NaN     NaN  
82           NaN       NaN     NaN  
83           NaN       NaN     NaN  
84           NaN       NaN     NaN  

[5 rows x 52 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 52 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Fact            80 non-null     object
 1   Fact Note       28 non-null     object
 2   Alabama         65 non-null     object
 3   Alaska          65 non-null     object
 4   Arizona         65 non-null     object
 5   Arkansas        65 non-null     object
 6   California      65 non-null     object
 7   Colorado        65 non-null     object
 8   Connecticut     65 non-null     object
 9   Delaware        65 non-null     object
 10  Florida         65 non-null     object
 11  Georgia         65 non-null     object
 12  Hawaii          65 non-null     object
 13  Idaho           65 non-null     object
 14  Illinois        65 non-null     object
 15  Indiana         65 non-null     object
 16  Iowa            65 non-null     object
 17  Kansas          65 non-null     object
 18  Kentucky    

None

>This dataset (Census dataset) has the Shape of (85, 52), it contains 85 instances and 52 features. These features are are objects of datatypes. It is noticted that these fetaurs have various features scales some of which are counts and some are percentegaes (%). From the initial look, the non-null value counts for each column has diffrent values than the shape of the data which has to be (85), most of them are missing (20) values, that indicated the missing values. Another indication for the missing valeus, is when looking at the tail of the datset, the NaN values are observed. For the memory usage of this dataset it is 34.7+ KB.

>The states data are shared between the two datasets, this will be explored in the bext section.

- Number of Unique values for each column in the dataset to make better optimization decisions.

In [269]:
census_dataset.nunique()

Fact              80
Fact Note         15
Alabama           65
Alaska            64
Arizona           64
Arkansas          64
California        63
Colorado          64
Connecticut       63
Delaware          64
Florida           64
Georgia           64
Hawaii            63
Idaho             63
Illinois          65
Indiana           63
Iowa              65
Kansas            65
Kentucky          65
Louisiana         65
Maine             64
Maryland          64
Massachusetts     65
Michigan          64
Minnesota         65
Mississippi       65
Missouri          64
Montana           65
Nebraska          65
Nevada            63
New Hampshire     65
New Jersey        65
New Mexico        64
New York          65
North Carolina    65
North Dakota      63
Ohio              64
Oklahoma          64
Oregon            63
Pennsylvania      65
Rhode Island      64
South Carolina    65
South Dakota      65
Tennessee         64
Texas             64
Utah              64
Vermont           63
Virginia     

> This

- Statics describing of the numerical features in the dataset:

In [272]:
census_dataset.describe(include='all').T

count unique                                           top freq
Fact              80     80  Population estimates, July 1, 2016,  (V2016)    1
Fact Note         28     15                                           (c)    6
Alabama           65     65                                     4,863,300    1
Alaska            65     64                                         7.30%    2
Arizona           65     64                                        50.30%    2
Arkansas          65     64                                        50.90%    2
California        65     63                                         6.80%    2
Colorado          65     64                                         3.30%    2
Connecticut       65     63                                         0.10%    2
Delaware          65     64                                        51.60%    2
Florida           65     64                                        51.10%    2
Georgia           65     64                                         6.40%    2
Hawaii            65     63                                     1,360,301    2
Idaho             65     63                                        49.90%    2
Illinois          65     65                                    12,801,539    1
Indiana           65     63                                         2.20%    2
Iowa              65     65                                     3,134,693    1
Kansas            65     65                                     2,907,289    1
Kentucky          65     65                                     4,436,974    1
Louisiana         65     65                                     4,681,666    1
Maine             65     64                                         1.60%    2
Maryland          65     64                                        51.60%    2
Massachusetts     65     65                                     6,811,779    1
Michigan          65     64                                         6.30%    2
Minnesota         65     65                                     5,519,952    1
Mississippi       65     65                                     2,988,726    1
Missouri          65     64                                        14.00%    2
Montana           65     65                                     1,042,520    1
Nebraska          65     65                                     1,907,116    1
Nevada            65     63                                        23.00%    2
New Hampshire     65     65                                     1,334,795    1
New Jersey        65     65                                     8,944,469    1
New Mexico        65     64                                         0.025    2
New York          65     65                                      19745289    1
North Carolina    65     65                                      10146788    1
North Dakota      65     63                                        672591    2
Ohio              65     64                                         0.022    2
Oklahoma          65     64                                         0.505    2
Oregon            65     63                                         0.038    2
Pennsylvania      65     65                                      12784227    1
Rhode Island      65     64                                          0.01    2
South Carolina    65     65                                       4961119    1
South Dakota      65     65                                        865454    1
Tennessee         65     64                                         0.048    2
Texas             65     64                                        50.40%    2
Utah              65     64                                         2.50%    2
Vermont           65     63                                         1.90%    2
Virginia          65     65                                     8,411,808    1
Washington        65     65                                     7,288,000    1
West Virginia     65     64                                      

>It is niticed that, 

### **Assess datasets**
In this sub-section, each of the datasets are assessed against having the follwing:
- Detecting Missing values and the percebtages of missing values.
- Detecting Duplicates values and the percentages of duplicates.
- Detecting outliers and the percentages of outliers.
- Detecting inconsistency of columns or values that would affect the mergere later on.

#### **Missing values detection:**
Let's detect wether there are missing values and what is the percentage of missing values in each of the datasets:

In [288]:
def find_missing_percentage(df, col=False, col_name=None):
    """
    To caculate the percentage of missing data in a column, or for the whole dataset
    Args:
        df (pd.DataFrame): specific data type that contains the col
        col (str): column name or key
    Returns:
        (float %): percentage of missing data
        KeyError if the column name is missing
    """
    if isinstance(df, pd.DataFrame):
        if col is True:
            ## find missing values percentage for specific column
            if col_name in df.columns:
                return (1 - (df[col].shape[0] - df[col].isna().sum())/np.prod(df[col].shape)) *100
            else:
                return "Key Error: missing column"
        else:
            ## find the missing values percentage in all dataset
            print("Total of missing percentage (%): ", (df.isna().sum().sum() / np.prod(df.shape)) * 100)
            return (df.isna().sum()/ np.prod(df.shape)) * 100
    else:
        print("TypeError: This is not a pandas DataFrame")

##### Guns Dataset
For this dataset, we know already that there are missing values from the initial exploration of the dataset. Let's take a look on the percentage of missing values to help in figuring out what is best to handle them:

In [286]:
find_missing_percentage(guns_dataset)

Total of missing percentage (%):  45.860959076818105


month                        0.000000
state                        0.000000
permit                       0.007120
permit_recheck               3.377386
handgun                      0.005933
long_gun                     0.005636
other                        2.072116
multiple                     0.000000
admin                        0.006823
prepawn_handgun              0.576395
prepawn_long_gun             0.576989
prepawn_other                2.186327
redemption_handgun           0.575505
redemption_long_gun          0.575802
redemption_other             2.186327
returned_handgun             3.051069
returned_long_gun            3.067385
returned_other               3.165280
rentals_handgun              3.410018
rentals_long_gun             3.458966
private_sale_handgun         2.887910
private_sale_long_gun        2.887910
private_sale_other           2.887910
return_to_seller_handgun     2.969489
return_to_seller_long_gun    2.887910
return_to_seller_other       3.034753
totals      

> It is 

##### Census Dataset

In [287]:
find_missing_percentage(census_dataset)

Total of missing percentage (%):  24.02714932126697


Fact              0.113122
Fact Note         1.289593
Alabama           0.452489
Alaska            0.452489
Arizona           0.452489
Arkansas          0.452489
California        0.452489
Colorado          0.452489
Connecticut       0.452489
Delaware          0.452489
Florida           0.452489
Georgia           0.452489
Hawaii            0.452489
Idaho             0.452489
Illinois          0.452489
Indiana           0.452489
Iowa              0.452489
Kansas            0.452489
Kentucky          0.452489
Louisiana         0.452489
Maine             0.452489
Maryland          0.452489
Massachusetts     0.452489
Michigan          0.452489
Minnesota         0.452489
Mississippi       0.452489
Missouri          0.452489
Montana           0.452489
Nebraska          0.452489
Nevada            0.452489
New Hampshire     0.452489
New Jersey        0.452489
New Mexico        0.452489
New York          0.452489
North Carolina    0.452489
North Dakota      0.452489
Ohio              0.452489
O

> It is 

#### **Duplicates detection:**

In [290]:
def find_duplicates_percentage(df, col =False, col_name = None):
    """
    To caculate the percentage of duplicated data in a column, or for the whole dataset
    Args:
        df (pd.DataFrame): specific data type that contains the col
        col (str): column name or key
    Returns:
        (float %): percentage of duplicted data
        KeyError if the column name is duplicted
    """
    if isinstance(df, pd.DataFrame):
        if col is True:
            ## find missing values percentage for specific column
            if col_name in df.columns:
                return (1 - (df[col].shape[0] - df[col].isna().sum())/np.prod(df[col].shape)) *100
            else:
                return "Key Error: missing column"
        else:
            ## find the missing values percentage in all dataset
            print("Total of duplicated percentage (%): ", (df.isna().sum().sum() / np.prod(df.shape)) * 100)
            return (df.isna().sum()/ np.prod(df.shape)) * 100
    else:
        print("TypeError: This is not a pandas DataFrame")
    

##### Guns Dataset

In [296]:
census_dataset.duplicated().sum()

3

> It is 

##### Census Dataset

> It is 

#### **Outliers detection:**

##### Guns Dataset

> It is 

##### Census Dataset

> It is 

#### **Inconsistency detection:**

The following states are not in the second dataset, but in the first one: 

In [248]:
set(dfs[0]['state'].unique()) - set(dfs[1].columns[2:])

{'District of Columbia',
 'Guam',
 'Mariana Islands',
 'Puerto Rico',
 'Virgin Islands'}

### Data Cleaning
In the previous section, the structure of the data and any problems that needed to be handeled were discussed and detected. In this sections, decisions are made about how to handle these problems to finally reach the best output.

- Mergeing the dataset into one:

In [ ]:
# 
#   cleaned, perform those cleaning steps in the second part of this section.

Spliting by comma to see what information the column Fact give:

In [ ]:
for i in range(85):
    if not isinstance(dfs[1]['Fact'][i], float):
        print(dfs[1]['Fact'][i].split(',')) 

['Population estimates', ' July 1', ' 2016', '  (V2016)']
['Population estimates base', ' April 1', ' 2010', '  (V2016)']
['Population', ' percent change - April 1', ' 2010 (estimates base) to July 1', ' 2016', '  (V2016)']
['Population', ' Census', ' April 1', ' 2010']
['Persons under 5 years', ' percent', ' July 1', ' 2016', '  (V2016)']
['Persons under 5 years', ' percent', ' April 1', ' 2010']
['Persons under 18 years', ' percent', ' July 1', ' 2016', '  (V2016)']
['Persons under 18 years', ' percent', ' April 1', ' 2010']
['Persons 65 years and over', ' percent', '  July 1', ' 2016', '  (V2016)']
['Persons 65 years and over', ' percent', ' April 1', ' 2010']
['Female persons', ' percent', '  July 1', ' 2016', '  (V2016)']
['Female persons', ' percent', ' April 1', ' 2010']
['White alone', ' percent', ' July 1', ' 2016', '  (V2016)']
['Black or African American alone', ' percent', ' July 1', ' 2016', '  (V2016)']
['American Indian and Alaska Native alone', ' percent', ' July 1', ' 

<a id='eda'></a>
## Exploratory Data Analysis

- **Compute statistics**: You should compute the relevant statistics throughout the analysis when an inference is made about the data.
- **create visualizations**:  Note that at least two or more kinds of plots should be created as part of the exploration, and you must  compare and show trends in the varied visualizations. Remember to utilize the visualizations that the pandas library already has available.

with the goal of addressing the research questions that you posed in the Introduction section. 

- Investigate the stated question(s) from multiple angles. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.
-  You should explore at least three variables in relation to the primary question. 
- This can be an exploratory relationship between three variables of interest, or looking at how two independent variables relate to a single dependent variable of interest. 
- Lastly, you  should perform both single-variable (1d) and multiple-variable (2d) explorations.


### Research Question 1 (Replace this header name!)

In [ ]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.

### Research Question 2  (Replace this header name!)

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

- Summarize the results accurately, and point out where additional research can be done or where additional information could be useful.
-  Make sure that you are clear with regards to the limitations of your exploration. You should have at least 1 limitation explained clearly. 
-  If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!
- cheack all the areas of the rubric. 

create a .html or .pdf version of this notebook in the workspace here. 

In [ ]:
# Running this cell will execute a bash command to convert this notebook to an .html file
!python -m nbconvert --to html Investigate_a_Dataset.ipynb